src/accelerate/state.py

```diff
@@ -143,231 +144,118 @@ def __init__(self, cpu: bool = False, **kwargs):
-                else:
-                    self.distributed_type = DistributedType.MULTI_CPU
-                # Actually, CCL_WORKER_COUNT is a CPU only env var in CCL, no need to set it for XPU.
-                if is_ccl_available() and (
-                    get_int_from_env(["CCL_WORKER_COUNT"], 0) > 0 or self.distributed_type == DistributedType.MULTI_XPU
-                ):
-                    if get_ccl_version() >= "1.12":
-                        import oneccl_bindings_for_pytorch  # noqa: F401
-                    else:
-                        import torch_ccl  # noqa: F401
-                    backend = "ccl"
-                elif torch.distributed.is_mpi_available():
-                    backend = "mpi"
-                else:
-                    backend = "gloo"
-                # Try to get launch configuration from environment variables set by MPI launcher - works for Intel MPI, OpenMPI and MVAPICH
-                rank = get_int_from_env(["RANK", "PMI_RANK", "OMPI_COMM_WORLD_RANK", "MV2_COMM_WORLD_RANK"], 0)
-                size = get_int_from_env(["WORLD_SIZE", "PMI_SIZE", "OMPI_COMM_WORLD_SIZE", "MV2_COMM_WORLD_SIZE"], 1)
-                local_rank = get_int_from_env(
-                    ["LOCAL_RANK", "MPI_LOCALRANKID", "OMPI_COMM_WORLD_LOCAL_RANK", "MV2_COMM_WORLD_LOCAL_RANK"], 0
-                )
-                local_size = get_int_from_env(
-                    ["LOCAL_WORLD_SIZE", "MPI_LOCALNRANKS", "OMPI_COMM_WORLD_LOCAL_SIZE", "MV2_COMM_WORLD_LOCAL_SIZE"],
-                    1,
-                )
-                self.local_process_index = local_rank
-                os.environ["RANK"] = str(rank)
-                os.environ["WORLD_SIZE"] = str(size)
-                os.environ["LOCAL_RANK"] = str(local_rank)
-                os.environ["LOCAL_WORLD_SIZE"] = str(local_size)
-
-                if backend == "ccl" and self.distributed_type == DistributedType.MULTI_XPU:
+                    if int(os.environ.get("LOCAL_RANK", -1)) != -1:
+                        if os.environ.get("ACCELERATE_USE_DEEPSPEED", "false") == "true":
+                            if not is_deepspeed_available():
+                                raise ImportError(
+                                    "DeepSpeed is not available => install it using `pip3 install deepspeed` or build it from source"
+                                )
+                            from deepspeed import comm as dist
+
+                            if is_xpu_available() and is_ccl_available():
+                                os.environ["CCL_PROCESS_LAUNCHER"] = "none"
+                                os.environ["CCL_LOCAL_SIZE"] = os.environ.get("LOCAL_WORLD_SIZE", "1")
+                                os.environ["CCL_LOCAL_RANK"] = os.environ.get("LOCAL_RANK", "0")
+                            dist.init_distributed(dist_backend=self.backend, auto_mpi_discovery=False, **kwargs)
+                            self.set_device()
```

Comment:
Just want to ensure that this `set_device()` call is correct here, since it is called again further down below. If it's indeed needed here, maybe add a comment to explain?

Reply:
It is, we need it to happen before the distributed type is set to deepspeed (and if it's called again after the first time, nothing happens)